[Merged by Bors] - feat: simple reflections, words, lifting in Coxeter groups

trivial1711 opened the pull request:
Prove basic properties of simple reflections in a Coxeter group. Add definitions `CoxeterSystem.lift`, and `CoxeterSystem.wordProd` to `GroupTheory.Coxeter.Basic` and prove their basic properties.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
- [x] depends on: #11493
- [x] depends on: #11804
- [x] depends on: #11836
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

trivial1711 commented:
Hello, I'm currently working on simplifying the proofs and improving the API.

trivial1711 commented:
Hello, I'm done simplifying the proofs. The PR is dependent on #11493 now, though.

trivial1711 commented on Mathlib/GroupTheory/Coxeter/Basic.lean:
```diff
@@ -455,9 +495,219 @@ def H₄ : Matrix (Fin 4) (Fin 4) ℕ :=
+    π (i :: ω) = s i * π ω := by simp [wordProd]
+
+theorem wordProd_singleton (i : B) :
+    π ([i]) = s i := by simp [wordProd]
```
I put it all on one line, but I didn't add the "simp" because `π [i]` unfolds to `π (i :: nil)`, which is not in simp-normal form because of `wordProd_cons`

trivial1711 commented on Mathlib/GroupTheory/Coxeter/Basic.lean:
```diff
@@ -455,9 +495,219 @@ def H₄ : Matrix (Fin 4) (Fin 4) ℕ :=
+
+/-! ### Words -/
+/-- The product of the simple reflections of `W` corresponding to the indices in `ω`.-/
+def wordProd (ω : List B) : W := prod (map cs.simpleReflection ω)
```
Yes, this is an essential definition, especially when we start talking about the length of an element.

trivial1711 commented:
I realized while revising this pull request that I had left two different ways to get simple reflections of a Coxeter system: the `FunLike` instance and the function `simpleReflection`. I slightly modified the code so everything is written in terms of the `simpleReflection` function and there is no `FunLike` instance, because I don't think that the `FunLike` instance is very intuitive.

YaelDillies commented on Mathlib/GroupTheory/Coxeter/Basic.lean:
```diff
@@ -0,0 +1,695 @@
+
+theorem wordProd_concat (i : B) (ω : List B) : π (ω.concat i) = π ω * s i := by simp [wordProd]
+
+@[simp] theorem wordProd_append (ω ω' : List B) : π (ω ++ ω') = π ω * π ω' := by simp [wordProd]
```
This lemma is not "affine": There are variables (namely `cs`) with more occurrences on the RHS than on the RHS. This means that if `cs` is simplifiable then you could hit exponential blowup:
`π (a ++ b ++ c ++ d ++ e) → π a * π b * π c * π d * π e → simplified π a * simplified π b * simplified π c * simplified π d * simplified π e`.

It is therefore advisable for this not to be a simp lemma.

YaelDillies commented on Mathlib/GroupTheory/Coxeter/Basic.lean:
```diff
@@ -0,0 +1,695 @@
+
+@[simp] theorem wordProd_nil : π [] = 1 := by simp [wordProd]
+
+@[simp] theorem wordProd_cons (i : B) (ω : List B) : π (i :: ω) = s i * π ω := by simp [wordProd]
```
Similarly, this lemma is not affine and should probably not be simp

YaelDillies commented on Mathlib/GroupTheory/Coxeter/Basic.lean:
```diff
@@ -0,0 +1,695 @@
+
+@[simp] theorem wordProd_cons (i : B) (ω : List B) : π (i :: ω) = s i * π ω := by simp [wordProd]
+
+theorem wordProd_singleton (i : B) : π ([i]) = s i := by simp [wordProd]
```
In return, this one gets to be affine
```suggestion
@[simp] theorem wordProd_singleton (i : B) : π [i] = s i := by simp [wordProd]
```

YaelDillies reviewed (commented):
Could you do the file rename in a preliminary PR? You seem to have confused git since the diff here doesn't show the rename.

trivial1711 commented:
Thanks for the feedback, I'll take care of this tomorrow.

YaelDillies reviewed (commented):
So why exactly did you remove the `FunLike` instance? This seems strictly less ergonomic than before, but we can't see it because you used local notation for `cs.simpleReflection`. Regardless of whether it's a good idea or not, this change should definitely be in your PR description somewhere.

trivial1711 commented on Mathlib/GroupTheory/Coxeter/Basic.lean:
```diff
@@ -95,8 +126,14 @@ variable (M : Matrix B B ℕ)
+    (e : B ≃ B') (hM : M.IsCoxeter) : (Matrix.reindex e e M).IsCoxeter where
+  symmetric := by dsimp only [Matrix.IsSymm]; rw [Matrix.transpose_reindex, hM.symmetric]
+  diagonal := by intro b; dsimp [Matrix.reindex]; exact hM.diagonal (e.symm b)
+  off_diagonal := by intro i i' hii'; dsimp [Matrix.reindex]; apply hM.off_diagonal; aesop
```
Yes, many things would benefit from being renamed. In fact, I am just going to go ahead and do it.

trivial1711 commented:
I realized today that the names of a lot of the theorems in the original file `GroupTheory.Coxeter.Basic` should be changed, so I went ahead and did it. I changed the PR description to describe exactly what I changed.

trivial1711 commented:
I am going to split this into two pull requests to make it easier to review.

ocfnash commented on Mathlib/GroupTheory/Coxeter/Basic.lean:
```diff
@@ -166,4 +186,179 @@ theorem _root_.CoxeterMatrix.toCoxeterSystem_simple (M : CoxeterMatrix B) :
+  val_inv := pow_one (f i * f i) ▸ M.diagonal i ▸ hf i i
+  inv_val := pow_one (f i * f i) ▸ M.diagonal i ▸ hf i i
+
+/-- Extend the function `f : B → G` to a monoid homomorphism `f' : W → G` satisfying
+`f' (cs.simple i) = f i` for all `i`. -/
+def lift {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) : W →* G :=
+  MonoidHom.comp (Units.coeHom G) (cs.groupLift
+    (show ∀ i i', ((restrictUnit hf) i * (restrictUnit hf) i') ^ M i i' = 1 from
+      fun i i' ↦ Units.ext (hf i i')))
+
+private theorem toMonoidHom_apply_symm_apply (a : PresentedGroup (M.relationsSet)):
+    (MulEquiv.toMonoidHom cs.mulEquiv : W →* PresentedGroup (M.relationsSet))
+    ((MulEquiv.symm cs.mulEquiv) a) = a := calc
+  _ = cs.mulEquiv ((MulEquiv.symm cs.mulEquiv) a) := by rfl
+  _ = _                                           := by rw [MulEquiv.apply_symm_apply]
+
+theorem lift_apply_simple {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) (i : B) :
+    cs.lift hf (s i) = f i := by
+  dsimp only [simple, lift, groupLift, MonoidHom.comp_apply]
+  rw [← MonoidHom.toFun_eq_coe, toMonoidHom_apply_symm_apply]
+  simp
+  rfl
```
There's a stronger statement available here, right:
```suggestion
private theorem toMonoidHom_apply_symm_apply (a : PresentedGroup (M.relationsSet)) :
    (MulEquiv.toMonoidHom cs.mulEquiv : W →* PresentedGroup (M.relationsSet))
    ((MulEquiv.symm cs.mulEquiv) a) = a := calc
  _ = cs.mulEquiv ((MulEquiv.symm cs.mulEquiv) a) := by rfl
  _ = _                                           := by rw [MulEquiv.apply_symm_apply]

/-- Given a Coxeter group `W` together with a Coxeter system with matrix `M`, indexed by `B`, then
for any monoid `G`, `M`-liftable maps `B → G` are equivalent to monoid homs `W →* G`. -/
def liftEquiv {G : Type*} [Monoid G] : {f : B → G | IsLiftable M f} ≃ (W →* G) where
  toFun f := MonoidHom.comp (Units.coeHom G) (cs.groupLift
    (show ∀ i i', ((restrictUnit f.property) i * (restrictUnit f.property) i') ^ M i i' = 1 from
      fun i i' ↦ Units.ext (f.property i i')))
  invFun f := ⟨f ∘ cs.simple, fun i i' ↦ by
    rw [comp_apply, comp_apply, ← map_mul, ← map_pow, simple_mul_simple_pow, map_one]⟩
  left_inv f := by
    ext i
    simp only [MonoidHom.comp_apply, comp_apply, mem_setOf_eq, groupLift, simple]
    rw [← MonoidHom.toFun_eq_coe, toMonoidHom_apply_symm_apply, PresentedGroup.toGroup.of,
      OneHom.toFun_eq_coe, MonoidHom.toOneHom_coe, Units.coeHom_apply]
    rfl
  right_inv f := by
    -- Some work to do here
    sorry
```
Unfortunately I have very little time now so I cannot do more than leave this brief comment.

ocfnash replied:
There's no need to have both `lift` and `liftEquiv` though, right? I mean it's just another syntactic stumbling block for a future `rw` if we have both isn't it?

trivial1711 replied:
How about I make it an isomorphism but still call it `lift`?

joelriou commented on Mathlib/GroupTheory/Coxeter/Basic.lean:
```diff
@@ -166,4 +186,179 @@ theorem _root_.CoxeterMatrix.toCoxeterSystem_simple (M : CoxeterMatrix B) :
+    Prop := ∀ i i', (f i * f i') ^ M i i' = 1
+
+private theorem relations_liftable {G : Type*} [Group G] {f : B → G} (hf : IsLiftable M f)
+  (r : FreeGroup B) (hr : r ∈ M.relationsSet) : (FreeGroup.lift f) r = 1 := by
```
```suggestion
    (r : FreeGroup B) (hr : r ∈ M.relationsSet) : (FreeGroup.lift f) r = 1 := by
```

joelriou commented on Mathlib/GroupTheory/Coxeter/Basic.lean:
```diff
@@ -293,30 +298,37 @@ private def restrictUnit {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M
+      PresentedGroup.toGroup.of, CoxeterSystem.restrictUnit, Units.coeHom_apply]
+    rfl
+
+theorem lift_apply_simple {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) :
+    ∀ i, cs.lift ⟨f, hf⟩ (s i) = f i := congrFun (congrArg Subtype.val (cs.lift.left_inv ⟨f, hf⟩))
```
The convention in mathlib is to put the variable `i` at the left of `:`, unless there are very good reasons not to do so. Then, why not making this a `simp` lemma?

trivial1711 replied:
Thanks, I've changed it.

joelriou commented:
Thanks!

bors merge
